### PR TITLE
zfs file system compatibility checks added. #67

### PR DESCRIPTION
**Pull request by dellgreen**
This is my first draft implementation. It has been tested on python 2.7 and 3.4.
All tests have been ran on a zfs file system (with and without fix applied) and an ext4 system.

### CODE REVIEW

**Comment by dellgreen** [comment]
Related to: #65

**Review by dedekind — APPROVED** [review]
This patch looks OK to me, thanks!

File: bmaptools/BmapHelpers.py [hunk 1]
```diff
@@ -83,3 +96,54 @@ def program_is_available(name):
...
+
+def get_file_system_type(filepath):
+    """
+    Return the file system type the file object file resides on.
+    """
+
+    # normalize,resolve and get absolute path to the file object
+    f = os.path.realpath(filepath)
+    cmd = "df -T '%s' | awk 'NR==2 {print $2}'" % f
+    proc = subprocess.Popen(cmd, stdout=PIPE, stderr=PIPE, shell=True, universal_newlines=True)
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:109]: Is it possible to avoid running external programs for this? Less of this is generally better, less dependencies. Did you look at other solutions, like just scanning /proc/mounds, or using psutils, here is some helpful stackoverflow discussion: https://stackoverflow.com/questions/7718411/determine-device-of-filesystem-in-python?lq=1

File: requirements-test.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+six
+backports.tempfile
+nose
+mock
```
Comment (dedekind) [inline requirements-test.txt:4]: This is not something needed for the new test, right? It is just something missing. If so, would be great to have it as a separate commit.

Then, I suggest to aggregate all the trivial patches that are sort of independent of this ZFS check into a separate pull request (this one, the .gitignore change, the first commit which cleans up the test). Then we could go ahead and merge those quickly and independently.

File: .gitignore [hunk 3]
```diff
@@ -59,3 +59,7 @@ docs/_build/
 # virtualenv
 venv/
 ENV/
+
+#vscode
+.vscode/
+.noseids
```
Comment (dedekind) [inline .gitignore:65]: Would be great to have this as independent commit.

**Review by dedekind — CHANGES_REQUESTED** [review]

File: bmaptools/BmapHelpers.py (hunk 1 above)
Comment (dellgreen) [reply bmaptools/BmapHelpers.py:109]: whilst psutil looks like a good approach, it looks like it comes with at a heavy price, the user would need to install gcc and python3-dev on linux and Xcode on mac.  Am i correct in thinking mac support is not a thing yet for bmap-tools and hence I dont need to worry about it?

Comment (dedekind) [reply bmaptools/BmapHelpers.py:109]: Probably we do not want psutil. I never worried about Windows or Mac.

I guess 'df' is OK, but please, do not pipe to awk, python has enough tools for parsing strings. I think all you need will be output.split()[0].

Also, please handle errors. Check how / if they are handled in different places in this project. I went through evolution of how I handle exception in my programs, and I do not remember how I did it in this project, but just be consistent with the rest of the code.

Comment (dellgreen) [reply bmaptools/BmapHelpers.py:109]: if mac support is required then, "df -T '%s' | awk 'NR==2 {print $2}" can be replaced with a single call to "mount" and the output of this would need to be parsed and correlated with a python function to determine the mount point that the image file resides on

Comment (smcv) [reply bmaptools/BmapHelpers.py:109]: > `cmd = "df -T '%s' | awk 'NR==2 {print $2}'" % f`

This is vulnerable to [shell injection](https://xkcd.com/327/). If the filename is `'$(rm -fr /)` (including the single quote, to escape from the single quotes in the format string) then running bmaptool will remove all your files. In general, avoid using `shell=True` with `subprocess`.

The best and most direct way to get this information would be `os.statvfs(f).f_type` and comparing with the filesystem's magic number, but Python doesn't seem to implement that struct field...

Using `subprocess.Popen(['stat', '-f', '--printf=%t'], ...)` (without `shell=True`), and comparing with the magic number of a ZFS filesystem (possibly 0x2f52f5?), might be a less bad way to do this with external commands; or maybe `%T` instead of `%t` and comparing with `zfs`, or similar.

Comment (dellgreen) [reply bmaptools/BmapHelpers.py:109]: looks a better way to go

Comment (dellgreen) [reply bmaptools/BmapHelpers.py:109]: a quick test, however is resulting in stat output not accurately  identifying the filesystem type being used. testing on files on an ext4 file system results in ext2/ext3 being returned for stat whereas df returns ext4.

Comment (smcv) [reply bmaptools/BmapHelpers.py:109]: ext2, ext3 and ext4 all have the same magic number, so it's expected that different tools will label them differently. What matters for this PR is surely whether ZFS can be detected?

Comment (dellgreen) [reply bmaptools/BmapHelpers.py:109]: given that these are support helper functions, I was aiming for this particular function to be accurate and reusable in-case the file system needs to be known for some other reasons in the future.

Comment (dellgreen) [reply bmaptools/BmapHelpers.py:109]: shell=true should no longer be needed with the removal of the pipe as per @dedekind request.

Comment (dellgreen) [reply bmaptools/BmapHelpers.py:109]: taking the best of @smcv  @dedekind recommendations it should be possible to do something like:

subprocess.Popen(['df', '-T', '--', f], ...)

and the process the output in python

**Comment by dellgreen** [comment]
latest refactoring now includes requested changes

File: bmaptools/BmapHelpers.py [hunk 4]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
             return True
 
     return False
+
+
+def get_mount_point(filepath):
+    """
+    Return the mount point the file object file resides on.
+    """
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:102]: The convention is to have a newline between the docstring and the code. Also for sort comments the quotes can be on the same line.

File: bmaptools/BmapHelpers.py [hunk 5]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
             return True
 
     return False
+
+
+def get_mount_point(filepath):
+    """
+    Return the mount point the file object file resides on.
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:101]: "the file object file" - not very well phrased. The input is path, the output is the mount point. I suggest:

Return the mount point for path in 'filepath'.

Also, there is not reason to limit this to files, it'll work for non-files too, right? Then just call the variable 'path'.

File: bmaptools/BmapHelpers.py [hunk 6]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+    """
+    f = os.path.realpath(filepath)
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:103]: Pylint complains about one-letter variables. Call it "part" or something, something along the lines of "partial path", or "path prefix", or "path component". But "f" is not a good choice for a variable name. Have it self-documenting, at least e little bit.

File: bmaptools/BmapHelpers.py [hunk 7]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+    while f != os.path.sep:
+        if os.path.ismount(f): return f
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:105]: Please, keep the return statement on a separate line. It may look unimportant, and it is indeed not a big deal, but I am a fan of consistency and I find code much easier to read when it is uniform. So let's mimic the existing code (and this is a general rule of thumb for contributions - mimic existing code, if willing to improve something, improve globally).

File: bmaptools/BmapHelpers.py [hunk 8]
```diff
@@ -37,6 +45,7 @@ def human_size(size):
 
     return "%.1f %s" % (size, 'EiB')
 
+
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:48]: Why do you add these newlines? I think the unspoken convention in this code is to have double newlines between classes and only single between functions. It may be silly, and not big deal, but it is what it is. It is better to be consistent, so the whole code base would look coherent and thus, easier to follow.

If you like 2 newlines, you can suggest a _separate_ pull request cleaning this up globally, not just in one file. All or nothing.

File: bmaptools/BmapHelpers.py [hunk 9]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+    f = os.path.realpath(filepath)
+    while f != os.path.sep:
+        if os.path.ismount(f): return f
+        f = os.path.realpath(os.path.join(f, os.pardir))
+    return f
+
+
+def get_file_system_type(filepath):
+    """
+    Return the file system type the file object file resides on.
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:112]: File object is a python object representing a file, say, something that open() returns. In your case you have just a string (more modern way is to use pathlib.Path and the like, but this codebase uses just strings). Try to improve the comment a bit, see the similar suggestion above.

File: bmaptools/BmapHelpers.py [hunk 10]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+    """
+
+    # normalize,resolve and get absolute path to the file object
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:115]: Just as a convention, start comments from a capital letter. A dot at the end is nice too, but I do not remember how strict I was about it in this code. Again, check how other comments are done and mimic the style.

File: bmaptools/BmapHelpers.py [hunk 11]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+    f = os.path.realpath(filepath)
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:116]: Similar feedback about the 'f' variable.

File: bmaptools/BmapHelpers.py (hunk 9 above)
Comment (dedekind) [reply bmaptools/BmapHelpers.py:112]: And similar feedback about 'filepath', make it 'path' if this function does not care if it is a file or not. If it cares, may be there needs to be a check?

File: bmaptools/BmapHelpers.py [hunk 12]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+    proc = subprocess.Popen(['df', '-T', '--', f], stdout=PIPE, stderr=PIPE)
+    stdout, stderr = proc.communicate()
+    ftype = None
+    if stdout:
+        t = stdout.splitlines()
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:121]: 't' is a bad name. Could you please improve it? E.g., use 'entries', or 'split_line', or something else that is logical and at least somewhat self-documenting. Also, it would be nice to add a small comment with an example of the output string that you are parsing here, makes the code easier to read. Right now I have no clue what is going on and I have to run that 'df -T' and check how the output looks like. A comment would help.

File: bmaptools/BmapHelpers.py [hunk 13]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+        if len(t) >= 2: 
+            t = t[1].split(None, 2)
+            if len(t) >= 2: 
+                ftype = t[1].lower()
+
+    if not ftype:
+        raise Error("no file system type was found: %s" % stderr)
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:128]: Surely the error message would be much more helpful if it actually included the _path_ for which the file-system type was not found?

File: bmaptools/BmapHelpers.py [hunk 14]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+    return ftype
+
+
+def get_zfs_compat_param_path():
+    """
+    Returns path to required zfs kernel compatibility param.
+    """
+
+    return '/sys/module/zfs/parameters/zfs_dmu_offset_next_sync'
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:137]: Remove the function please, it brings no value. If this path is used only once, have a local variable. Otherwise, introduce a global variable instead.

File: bmaptools/BmapHelpers.py [hunk 15]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+    Return the mount point the file object file resides on.
+    """
+    f = os.path.realpath(filepath)
+    while f != os.path.sep:
+        if os.path.ismount(f): return f
+        f = os.path.realpath(os.path.join(f, os.pardir))
+    return f
+
+
+def get_file_system_type(filepath):
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:110]: FYI, in python we use leading underscore to communicate that a global symbol is private to this module and should not be used outside of the module. I do not know if this entire project is careful about it, but at least for this and other functions you add, mark them as private by using this name: _get_file_system_type().Those functions that you do want to be visible outside of this module (BmapHelpers) should start with a non-_.

File: bmaptools/BmapHelpers.py [hunk 16]
```diff
@@ -83,3 +94,72 @@ def program_is_available(name):
...
+
+    return '/sys/module/zfs/parameters/zfs_dmu_offset_next_sync'
+
+
+def is_zfs_compatible():
+    """
+    Returns whether hosts zfs configuration is compatible for our usage.
+    """
+
+    fp = get_zfs_compat_param_path()
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:145]: I guess 'fp' is file path? Surely this is poor name, because it is not just a file path, this is a path to a specific ZFS config file, and name of the variable could be improved, right? Even 'cfg_path' is much better. Or 'zfs_param_path' or 'param_path' or something.

**Review by dedekind — COMMENTED** [review]
Did not review everything, but done partial review. Mostly nitpicks and stylistical, but it is improtant to keep the code consistent in my opinion. Would you please consider addressing those nitpicks? Thanks!

File: bmaptools/BmapHelpers.py [hunk 17]
```diff
@@ -95,70 +93,61 @@ def program_is_available(name):
 
     return False
 
+def get_mount_point(path):
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:96]: This function is marked as public (does not start with "_"), is this intentional? I did not spot the user outside this module.

File: bmaptools/BmapHelpers.py [hunk 18]
```diff
@@ -95,70 +93,61 @@ def program_is_available(name):
...
-        f = os.path.realpath(os.path.join(f, os.pardir))
-    return f
+    part = os.path.realpath(path)
+    while part != os.path.sep:
+        if os.path.ismount(part): 
+            return part
+        part = os.path.realpath(os.path.join(part, os.pardir))
+    return part
 
+def get_file_system_type(path):
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:106]: Ditto (private vs public function).

File: bmaptools/BmapHelpers.py [hunk 19]
```diff
@@ -95,70 +93,61 @@ def program_is_available(name):
...
 
-def get_file_system_type(filepath):
-    """
-    Return the file system type the file object file resides on.
-    """
-
-    # normalize,resolve and get absolute path to the file object
-    f = os.path.realpath(filepath)
-    proc = subprocess.Popen(['df', '-T', '--', f], stdout=PIPE, stderr=PIPE)
+    # Normalize,resolve and get absolute path of the supplied path.
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:109]: I find this comment a bit confusing - is it explaining what's 'os.path.realpath()' ? If so, I spent more time reading it and trying to understand the conveyed message. Let's remove it in that case. If I misunderstood it, may be we can improve it?

File: bmaptools/BmapHelpers.py [hunk 20]
```diff
@@ -95,70 +93,61 @@ def program_is_available(name):
...
+    abspath = os.path.realpath(path)
+    proc = subprocess.Popen(['df', '-T', '--', abspath], stdout=PIPE, stderr=PIPE)
     stdout, stderr = proc.communicate()
+
+    # Parse the output of subprocess, for example:
+    # Filesystem                Type 1K-blocks     Used Available Use% Mounted on
+    # rpool/USERDATA/foo_5ucog2 zfs  456499712 86956288 369543424  20% /home/foo
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:116]: Very helpful, thanks. But this comment basically explains the block of code below, so make it be adjacent to that block, remove the empty newline below.

File: bmaptools/BmapHelpers.py [hunk 21]
```diff
@@ -95,70 +93,61 @@ def program_is_available(name):
...
+
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:117]: This newline should go I think. Not big deal, but improves the readability.

File: bmaptools/BmapHelpers.py [hunk 22]
```diff
@@ -95,70 +93,61 @@ def program_is_available(name):
...
-                ftype = t[1].lower()
+        lines = stdout.splitlines()
+        if len(lines) >= 2: 
+            fields = lines[1].split(None, 2)
+            if len(fields) >= 2: 
+                ftype = fields[1].lower()
 
     if not ftype:
-        raise Error("no file system type was found: %s" % stderr)
+        raise Error("no file system type was found for '%s': %s" % (path, stderr))
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:127]: I think it is better to include both stdout and stderr in the error message. Will be much more helpful in case this actually happen. How about something like this:

raise Error("failed to find file system type for path at '%s'.\nHere is the 'df -T' output.\nstdout:\n%s\nstderr:\n%s" % (path, stdout, stderr))

File: bmaptools/BmapHelpers.py [hunk 23]
```diff
@@ -95,70 +93,61 @@ def program_is_available(name):
...
-    """
-
-    return '/sys/module/zfs/parameters/zfs_dmu_offset_next_sync'
-
 
 def is_zfs_compatible():
-    """
-    Returns whether hosts zfs configuration is compatible for our usage.
-    """
+    """Return if hosts zfs configuration is compatible."""
```
Comment (dedekind) [inline bmaptools/BmapHelpers.py:133]: Naming: is_zfs_compatible sounds a bit like "is ZFS-compatible", as if we were checking if something is compatible with ZFS.

May be this name would be a tiny bit better: is_zfs_configuration_compatible() ? (or config for brevity).

**Review by dedekind — COMMENTED** [review]
Thank you! I've done some more partial review and suggested some more improvements.

Is it normal practice in github to address review comments with an incremental patch? In my practice people just rework the patch and re-submit the whole thing instead. Why project history should include all the revisions with all the issues fixed? Who'd be interested in that?

Unless you have a good reason for addressing every review cycle with a separate patch, please, squash the changes. Thanks!

File: bmaptools/BmapHelpers.py (hunk 17 above)
Comment (dellgreen) [reply bmaptools/BmapHelpers.py:96]: actually, this method is not actually used so this can be removed

File: bmaptools/BmapHelpers.py (hunk 18 above)
Comment (dellgreen) [reply bmaptools/BmapHelpers.py:106]: given that this is a helpers module I fell this particular method could be quite useful going forward for any future work that may need to know the file system type of a path, even though its only currently used internally. If you would prefer it to be private I can change it.

**Comment by dellgreen** [comment]
> Is it normal practice in github to address review comments with an incremental patch? In my practice people just rework the patch and re-submit the whole thing instead. Why project history should include all the revisions with all the issues fixed? Who'd be interested in that?

Yes it is normal on larger Guthub projects, alot of them i have worked on insist on this method and only request a squash once all changes have been approved or have bots that auto-squash and commit them.
I personally have no preference, and are more than happy to squash these each time for you. :)

**Comment by dedekind** [comment]
Some more stylistic nitpicks, otherwise looks good to me, thanks. Please, just to make sure, do run it on non-ZFS to verify that there are no regressions. Also I assume you tested it on both "correctly" configured ZFS and "incorrectly" configured ZFS (correctly/incorrectly from bmap-tools point of view). Am I right in my assumption? Thank you!

**Comment by dedekind** [comment]
Hi Dell, would you mind to do those final polishing things and we can merge this one?

**Comment by dellgreen** [comment]
> Hi Dell, would you mind to do those final polishing things and we can merge this one?

Hi still here, sorry, had higher priority work this week to attend to. I should be able to take a look at these at the weekend. :)

**Review by dedekind — APPROVED** [review]

**Comment by dedekind** [comment]
I spotted few minor things, but I think it is good enough. Could you please confirm that you did run pylint on the files you changed and addressed new complaints, and that you did test this patch on ZFS with and without the feature, as well as on non-ZFS.

If yes, we can merge it. If not, would you please do those things? Thank you very much!

**Comment by dellgreen** [comment]
> I spotted few minor things, but I think it is good enough. Could you please confirm that you did run pylint on the files you changed and addressed new complaints, and that you did test this patch on ZFS with and without the feature, as well as on non-ZFS.
> 
> If yes, we can merge it. If not, would you please do those things? Thank you very much!

1. ran pylint on new files and changed files, and fixed the pylint errors only related to these
2. patch has been tested on ZFS system **without** zfs param changed and process fails
3. patch has been tested on ZFS system **with** zfs param changed and process works
4. patch has been tested on non-ZFS system and process works

so we should be good to merge :)

**Comment by dedekind** [comment]
Thank you for the contribution!
